### PR TITLE
Allow semi-colon after record type constraints

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/45538

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -1833,6 +1833,7 @@ private TypeParameterConstraintClauseSyntax ParseTypeParameterConstraintClause()
                     while (true)
                     {
                         if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken
+                            || this.CurrentToken.Kind == SyntaxKind.SemicolonToken
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1836]: >|| this.CurrentToken.Kind == SyntaxKind.SemicolonToken [](start = 28, length = 54)

It feels like we already should be able to stop at ```;```. For example, we are able to parse ```public abstract void M<T>() where T: class;``` without an error.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1836]: It looks like this is achieved by taking advantage of ```TerminatorState.IsEndOfMethodSignature```. It feels like it would be more robust to follow the same approach, who knows where else we need to check for the terminator, and it looks like we already have an infrastructure to deal with the task.

---
In reply to: [454654156](https://github.com/dotnet/roslyn/pull/45943#discussion_r454654156) [](ancestors = 454654156)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1836]: We might be able to drop explicit checks for ```;``` elsewhere if they were added for the benefit of records parsing.

---
In reply to: [454659255](https://github.com/dotnet/roslyn/pull/45943#discussion_r454659255) [](ancestors = 454659255,454654156)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 2]
```diff
@@ -16383,5 +16383,35 @@ protected C(C<T> other)
...
+#nullable enable
+record R<T>(T P) where T : class;
+record R2<T>(T P) where T : class { }
+
+public class C
+{
+    public static void Main()
+    {
+        var r = new R<string?>(""R"");
+        var r2 = new R2<string?>(""R2"");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:16400]: >var r2 = new R2<string?>(""R2""); [](start = 8, length = 33)

Consider adding a separate test for a "hard" error. For example, ```R<int>```.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 3]
```diff
@@ -408,6 +409,91 @@ public void RecordParsingAmbiguities()
             EOF();
         }
 
+        [Fact, WorkItem(45538, "https://github.com/dotnet/roslyn/issues/45538")]
+        public void RecordParsing_ConstraintsAndSemiColon()
+        {
+            var tree = ParseTree("record R<T> where T : class;", options: TestOptions.RegularPreview);
+            tree.GetDiagnostics().Verify();
+
+            UsingNode((CSharpSyntaxNode)tree.GetRoot());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:418]: >UsingNode((CSharpSyntaxNode)tree.GetRoot()); [](start = 12, length = 44)

Wouldn't ```UsingTree``` helper work? Why use a different pattern to verify diagnostic and the tree?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1836]: Hm, I just realized a problem with this change. It allows `abstract class C<T> where T : class;`.
That's not allowed by the [spec](https://github.com/dotnet/csharplang/blob/master/spec/classes.md#class-body).

Should we allow for consistency, or keep disallowing?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1836]: > Should we allow for consistency, or keep disallowing?

I think the ```;``` should only terminate record.

---
In reply to: [454688376](https://github.com/dotnet/roslyn/pull/45943#discussion_r454688376) [](ancestors = 454688376)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: 1. I made the change to leverage `IsTerminator()` infrastructure in new commit, so that you could see what it looks like, but frankly I think it's not better. Let me know if okay to revert.
If we have an explicit exit here with `{`, then I think it's okay to have an explicit exit for `;` as well. Neither character can be inside a constraint clause.

2. An explicit check for `;` was added for records parsing in `ParseClassOrStructOrInterfaceDeclaration` at line 1477, but I don't think it could benefit from `IsTerminator()` infrastructure.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -1467,8 +1467,15 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
             {
                 if (this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword)
                 {
+                    var saveTerm2 = _termState;
+                    if (keyword.Kind == SyntaxKind.RecordKeyword)
+                    {
+                        _termState |= TerminatorState.IsEndOfMethodSignature;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1473]: >IsEndOfMethodSignature [](start = 54, length = 22)

1) I think we want to add special member for end of record header, perhaps with the same behavior, but specific for records.
2) I think we want to move adjustment for ```_termState``` to the beginning of this method. So that ```ParseTypeParameterList``` and ```ParseBaseList``` calls could take advantage of it as well. We can remove it right before ```if (!(keyword.Kind == SyntaxKind.RecordKeyword) || CurrentToken.Kind != SyntaxKind.SemicolonToken)```.
3) I think we want to remove explicit checks for semicolon from ```ParseBaseList``` and ```SkipBadBaseListTokens```. I believe those were added there specifically to support records and we should rely on existing infrastructure instead. That will also prevent us from affecting non-record declarations (non-critical, I believe, but if it comes for free, I would take this).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: > I made the change to leverage IsTerminator() infrastructure in new commit, so that you could see what it looks like, but frankly I think it's not better. Let me know if okay to revert.
If we have an explicit exit here with {, then I think it's okay to have an explicit exit for ; as well. Neither character can be inside a constraint clause.

I think the approach that relies on ```IsTerminator``` infrastructure is better, assuming we will be able to make changes that I suggested in ```ParseClassOrStructOrInterfaceDeclaration```.

> An explicit check for ; was added for records parsing in ParseClassOrStructOrInterfaceDeclaration at line 1477, but I don't think it could benefit from IsTerminator() infrastructure.

I agree with this, that check is already very specific for records and it makes sense to keep it explicit.

---
In reply to: [454699232](https://github.com/dotnet/roslyn/pull/45943#discussion_r454699232) [](ancestors = 454699232)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -1756,11 +1762,10 @@ private BaseListSyntax ParseBaseList(SyntaxToken typeKeyword, bool haveParameter
                 list.Add(argumentList is object ? _syntaxFactory.PrimaryConstructorBaseType(firstType, argumentList) : (BaseTypeSyntax)_syntaxFactory.SimpleBaseType(firstType));
 
                 // any additional types
-                while (true)
+                int lastTokenPosition = -1;
+                while (IsMakingProgress(ref lastTokenPosition))
                 {
-                    if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken ||
-                        this.CurrentToken.Kind == SyntaxKind.SemicolonToken ||
-                        this.IsCurrentTokenWhereOfConstraintClause())
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1763]: Note: I intentionally simplified `this.IsCurrentTokenWhereOfConstraintClause()` to `this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword`, to align the abort logic used before and after the success case.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 6]
```diff
@@ -1864,6 +1880,30 @@ private TypeParameterConstraintClauseSyntax ParseTypeParameterConstraintClause()
...
+        private bool NotExpectedForEndOfTypeConstraintClause()
+        {
+            return this.CurrentToken.Kind != SyntaxKind.CommaToken && !this.IsPossibleTypeParameterConstraint();
+        }
+
+        private static readonly Func<LanguageParser, bool> s_abortForEndOfTypeConstraintClause = p => p.AbortForEndOfTypeConstraintClause();
+        private bool AbortForEndOfTypeConstraintClause()
+        {
+            return this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword
+                   || this.CurrentToken.Kind == SyntaxKind.EqualsGreaterThanToken // ex: 'void localFunction<T>() where T : class => ...'
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1905]: Note: we could probably handle the `=>` as a terminator as well, but that seems out of scope for current PR.

**Comment by AlekseyTs** [comment]
>             _termState = saveTerm;

It feels like it would be better to leave ```IsEndOfTypeSignature``` in effect here than to do another change/restore around ```ParseTypeParameterConstraintClauses``` call. We can restore original state right before ```if (!(keyword.Kind == SyntaxKind.RecordKeyword) || CurrentToken.Kind != SyntaxKind.SemicolonToken)```.  

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1463 in 5c75fe4. [](commit_id = 5c75fe4c038dcfa95be7723181336133e03de7e8, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -1449,7 +1451,8 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
             var keyword = ConvertToKeyword(this.EatToken());
 
             var saveTerm = _termState;
-            _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop;
+            _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop | TerminatorState.IsEndOfTypeSignature;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1454]: >TerminatorState.IsEndOfTypeSignature [](start = 81, length = 36)

I don't think we want semicolon to affect non-record type declarations.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -1467,8 +1470,11 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
             {
                 if (this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword)
                 {
+                    var saveTerm2 = _termState;
+                    _termState |= TerminatorState.IsEndOfTypeSignature;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1474]: >IsEndOfTypeSignature [](start = 50, length = 20)

I don't think we want semicolon to affect non-record type declarations.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 9]
```diff
@@ -1756,11 +1762,10 @@ private BaseListSyntax ParseBaseList(SyntaxToken typeKeyword, bool haveParameter
                 list.Add(argumentList is object ? _syntaxFactory.PrimaryConstructorBaseType(firstType, argumentList) : (BaseTypeSyntax)_syntaxFactory.SimpleBaseType(firstType));
 
                 // any additional types
-                while (true)
+                int lastTokenPosition = -1;
+                while (IsMakingProgress(ref lastTokenPosition))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1766]: >IsMakingProgress(ref lastTokenPosition) [](start = 23, length = 39)

Could you please elaborate why this change is needed?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 10]
```diff
@@ -1782,14 +1787,26 @@ private BaseListSyntax ParseBaseList(SyntaxToken typeKeyword, bool haveParameter
...
+            {
+                return this.SkipBadSeparatedListTokensWithExpectedKind(ref colon, list,
+                    s_notExpectedForEndOfBaseList,
+                    s_abortForEndOfBaseList,
+                    expected);
+            }
         }
 
-        private PostSkipAction SkipBadBaseListTokens(ref SyntaxToken colon, SeparatedSyntaxListBuilder<BaseTypeSyntax> list, SyntaxKind expected)
+        private static readonly Func<LanguageParser, bool> s_notExpectedForEndOfBaseList = p => p.NotExpectedForEndOfBaseList();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1800]: >s_notExpectedForEndOfBaseList [](start = 59, length = 29)

Why is this necessary? It looks like there is only one use-site and the lambda didn't capture anything, wasn't it cached by the compiler?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 11]
```diff
@@ -1782,14 +1787,26 @@ private BaseListSyntax ParseBaseList(SyntaxToken typeKeyword, bool haveParameter
...
+        private bool NotExpectedForEndOfBaseList()
         {
-            return this.SkipBadSeparatedListTokensWithExpectedKind(ref colon, list,
-                p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleAttribute(),
-                p => p.CurrentToken.Kind == SyntaxKind.SemicolonToken || p.CurrentToken.Kind == SyntaxKind.OpenBraceToken || p.IsCurrentTokenWhereOfConstraintClause() || p.IsTerminator(),
-                expected);
+            return this.CurrentToken.Kind != SyntaxKind.CommaToken && !this.IsPossibleAttribute();
+        }
+
+        private static readonly Func<LanguageParser, bool> s_abortForEndOfBaseList = p => p.AbortForEndOfBaseList();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1806]: >s_abortForEndOfBaseList [](start = 59, length = 23)

Same question.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 12]
```diff
@@ -1782,14 +1787,26 @@ private BaseListSyntax ParseBaseList(SyntaxToken typeKeyword, bool haveParameter
...
+                return this.SkipBadSeparatedListTokensWithExpectedKind(ref colon, list,
+                    s_notExpectedForEndOfBaseList,
+                    s_abortForEndOfBaseList,
+                    expected);
+            }
         }
 
-        private PostSkipAction SkipBadBaseListTokens(ref SyntaxToken colon, SeparatedSyntaxListBuilder<BaseTypeSyntax> list, SyntaxKind expected)
+        private static readonly Func<LanguageParser, bool> s_notExpectedForEndOfBaseList = p => p.NotExpectedForEndOfBaseList();
+        private bool NotExpectedForEndOfBaseList()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1801]: >NotExpectedForEndOfBaseList [](start = 21, length = 27)

It looks like there is only one call site, the extraction of the method feels unnecessary.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 13]
```diff
@@ -1830,11 +1847,10 @@ private TypeParameterConstraintClauseSyntax ParseTypeParameterConstraintClause()
                     bounds.Add(this.ParseTypeParameterConstraint());
 
                     // remaining bounds
-                    while (true)
+                    int lastTokenPosition = -1;
+                    while (IsMakingProgress(ref lastTokenPosition))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1851]: >IsMakingProgress(ref lastTokenPosition) [](start = 27, length = 39)

Why is this change needed?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I filed a follow-up issue for delegate caching after confirming from IL: https://github.com/dotnet/roslyn/issues/46073

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 14]
```diff
@@ -408,6 +409,657 @@ public void RecordParsingAmbiguities()
             EOF();
         }
 
+        [Fact, WorkItem(45538, "https://github.com/dotnet/roslyn/issues/45538")]
+        public void RecordParsing_ConstraintAndSemiColon()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:413]: >RecordParsing_ConstraintAndSemiColon [](start = 20, length = 36)

:memo: by reverting my code change, I confirmed that the only three tests that are affected by the change are record scenarios:
- RecordParsing_ConstraintAndSemiColon
- RecordParsing_ConstraintAndSemiColon_MissingColon
- RecordParsing_TwoConstraintsAndSemiColon

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 15]
```diff
@@ -408,6 +409,657 @@ public void RecordParsingAmbiguities()
...
+                    M(SyntaxKind.OpenBraceToken);
+                    M(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        [Fact, WorkItem(45538, "https://github.com/dotnet/roslyn/issues/45538")]
+        public void AbstractMethod_ConstrainsAndSemiColon()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:676]: >Constrains [](start = 35, length = 10)

Typo

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 16]
```diff
@@ -408,6 +409,657 @@ public void RecordParsingAmbiguities()
...
+        {
+            UsingTree("abstract class C { abstract void M<T>() where T : class; }");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:678]: >abstract class C { abstract void M<T>() where T : class; } [](start = 23, length = 58)

Is this a new scenario? We must have many tests that hit this already.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 17]
```diff
@@ -408,6 +409,657 @@ public void RecordParsingAmbiguities()
...
+                        {
+                            N(SyntaxKind.ClassKeyword);
+                        }
+                    }
+                    N(SyntaxKind.SemicolonToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:494]: >} [](start = 8, length = 1)

Consider testing:
```C#
record R<T> where T : class, new();
record R<T> where T : class,;
```

**Comment by AlekseyTs** [comment]
>                 p => p.CurrentToken.Kind == SyntaxKind.SemicolonToken || p.CurrentToken.Kind == SyntaxKind.OpenBraceToken || p.IsCurrentTokenWhereOfConstraintClause() || p.IsTerminator(),

I think this check can be removed now.

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1802 in 03f0649. [](commit_id = 03f0649092fc1edc41d96c9b17727735edb8af65, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Looks good (iteration 11), other than https://github.com/dotnet/roslyn/pull/45943#issuecomment-660236716

**Comment by AlekseyTs** [comment]
>                 p => p.CurrentToken.Kind == SyntaxKind.SemicolonToken || p.CurrentToken.Kind == SyntaxKind.OpenBraceToken || p.IsCurrentTokenWhereOfConstraintClause() || p.IsTerminator(),

To clarify. This is regarding ```p.CurrentToken.Kind == SyntaxKind.SemicolonToken```

---
In reply to: [660236716](https://github.com/dotnet/roslyn/pull/45943#issuecomment-660236716) [](ancestors = 660236716)

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1802 in 03f0649. [](commit_id = 03f0649092fc1edc41d96c9b17727735edb8af65, deletion_comment = False)

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs (hunk 16 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:678]: Consider changing the test to `abstract record R { abstract void M<T>() where T : class; }`

---
In reply to: [456567211](https://github.com/dotnet/roslyn/pull/45943#discussion_r456567211) [](ancestors = 456567211)
